tdrs-backend/tdpservice/parsers/util.py

```diff
@@ -174,3 +174,67 @@ def get_years_apart(rpt_month_year_date, date):
     delta = rpt_month_year_date - date
     age = delta.days/365.25
     return age
+
+
+class SortedRecords:
+    """Maintains a dict sorted by hash_val and model_type."""
+
+    def __init__(self, section):
+        self.records_are_s1_or_s2 = section in {'A', 'C'}
+        self.hash_sorted_cases = dict()
+        self.cases = dict()
+        self.cases_already_removed = set()
+        self.serialized_cases = set()
+
+    def add_record(self, hash_val, record_doc_pair):
+        """Add a record_doc_pair to the sorted object if the case hasn't been removed already."""
+        record, document = record_doc_pair
+        rpt_month_year = str(getattr(record, 'RPT_MONTH_YEAR'))
+
+        if hash_val in self.cases_already_removed:
+            logger.info("Record's case has already been removed due to category four errors. Not adding record with "
+                        f"info: ({record.RecordType}, {getattr(record, 'CASE_NUMBER', None)}, {rpt_month_year})")
+
+        if hash_val is not None and hash_val not in self.cases_already_removed:
+            hashed_case = self.hash_sorted_cases.get(hash_val, {})
+            records = hashed_case.get(document, [])
+            records.append(record)
+
+            hashed_case[document] = records
+            self.hash_sorted_cases[hash_val] = hashed_case
+            # We treat the nested dictionary here as a set because dictionaries are sorted while sets aren't. If we
```

Comment:
If you mean dict is automatically sorts on keys, I don't think that is true

Reply:
As of Python 3.7 dictionaries are [ordered](https://docs.python.org/3.7/whatsnew/3.7.html#:~:text=the%20insertion%2Dorder%20preservation%20nature%20of%20dict%20objects%20has%20been%20declared%20to%20be%20an%20official%20part%20of%20the%20Python%20language%20spec.)